# Блок 7. Основы искусственного интеллекта и машинного обучения
## Занятие 7. Деревья решений (Decision Tree)

### Что такое дерево решений

Дерево решений — один из самых понятных алгоритмов машинного обучения.

Алгоритм задаёт последовательность вопросов:

- Доход больше 50 000?
- Возраст больше 30?
- Есть кредит?

И постепенно приходит к решению.

### Преимущества

- легко объяснять;
- легко визуализировать;
- подходит для бизнеса;
- показывает важность признаков.

### Недостатки

- может переобучаться;
- чувствителен к шуму.

### Новые понятия

- узел (node);
- лист (leaf);
- глубина дерева (max_depth);
- переобучение.

### Связь с дипломом

Деревья решений часто используют:

- HR;
- банки;
- недвижимость;
- продажи;
- медицину.

## Ячейка 1. Загрузка датасета

### Алгоритм
Используем Breast Cancer.

In [1]:
from sklearn.datasets import load_breast_cancer
import pandas as pd

data=load_breast_cancer()

X=pd.DataFrame(data.data,columns=data.feature_names)
y=pd.Series(data.target)

print(X.shape)

assert len(X)==len(y)

(569, 30)


## Ячейка 2. Изучение данных

### Алгоритм
Смотрим признаки.

In [2]:
display(X.head())

print(data.target_names)

assert X.shape[1] > 10

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


['malignant' 'benign']


## Ячейка 3. Train/Test Split

### Алгоритм
Разделяем данные.

In [3]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test=train_test_split(
    X,y,test_size=0.2,random_state=42,stratify=y
)

assert len(X_train)>len(X_test)

## Ячейка 4. Первое дерево решений

### Алгоритм
Обучаем DecisionTreeClassifier.

Инициализация модели: tree=DecisionTreeClassifier(random_state=42) создает экземпляр дерева решений. Параметр random_state=42 фиксирует генератор случайных чисел, чтобы результаты обучения были воспроизводимыми при каждом запуске.

Обучение модели: tree.fit(X_train, y_train) запускает процесс обучения дерева на тренировочных признаках X_train и соответствующих им метках классов y_train. На этом этапе алгоритм строит структуру ветвления (вопросов и ответов), которая наилучшим образом разделяет данные.

In [4]:
from sklearn.tree import DecisionTreeClassifier

tree=DecisionTreeClassifier(random_state=42)

tree.fit(X_train,y_train)

assert tree is not None

## Ячейка 5. Предсказания

### Алгоритм
Получаем прогноз.

In [5]:
pred=tree.predict(X_test)

print(pred[:10])

assert len(pred)==len(y_test)

[0 1 0 1 0 1 1 0 0 0]


## Ячейка 6. Метрики качества

### Алгоритм
Считаем Accuracy.

In [6]:
from sklearn.metrics import accuracy_score

accuracy=accuracy_score(y_test,pred)

print("Accuracy:",accuracy)

assert accuracy>0.8

Accuracy: 0.9122807017543859


## Ячейка 7. Глубина дерева

### Алгоритм
Изучаем сложность модели.

Этот код используется для анализа структуры обученного дерева решений tree:

depth = tree.get_depth(): Этот метод возвращает глубину дерева решений. Глубина — это максимальное расстояние от корня дерева (самого первого вопроса) до его самого удаленного листа (конечного узла с решением). В вашем случае глубина равна 7, то есть в самой длинной цепочке вопросов алгоритм совершил 7 последовательных разветвлений.

leaves = tree.get_n_leaves(): Этот метод возвращает количество листьев в дереве решений. Листья (leaves) — это конечные узлы дерева, в которых не задаются вопросы, а принимается окончательное решение (прогноз класса). В вашем случае дерево имеет 19 листьев, то есть существует 19 уникальных финальных правил/групп для классификации объектов.

In [7]:
depth=tree.get_depth()
leaves=tree.get_n_leaves()

print("Depth:",depth)
print("Leaves:",leaves)

assert depth>0

Depth: 7
Leaves: 19


## Ячейка 8. Важность признаков

### Алгоритм
Находим самые важные признаки.

Этот код используется для оценки и сортировки важности признаков, которые дерево решений использовало для принятия решений:

tree.feature_importances_: После обучения модели дерево решений присваивает каждому входному признаку коэффициент важности (число от 0 до 1). Чем чаще признак использовался для разделения данных и чем эффективнее было это разделение, тем выше этот показатель. Сумма всех важностей равна 1.

importance = pd.DataFrame({...}): Создает таблицу (DataFrame) из двух колонок:

"feature": названия признаков из вашего набора данных (X.columns).
"importance": соответствующие им коэффициенты важности из обученной модели (tree.feature_importances_).
importance = importance.sort_values(by="importance", ascending=False): Сортирует полученную таблицу по колонке важности (by="importance") в убывающем порядке (ascending=False). Это позволяет сразу увидеть наиболее влиятельные признаки в самом верху таблицы (например, признак worst radius оказался самым важным в вашей модели).

In [8]:
importance=pd.DataFrame({
    "feature":X.columns,
    "importance":tree.feature_importances_
})

importance=importance.sort_values(
    by="importance",
    ascending=False
)

display(importance.head(10))

assert len(importance)>0

,feature,importance
20,worst radius,0.697811
27,worst concave points,0.116083
21,worst texture,0.058881
11,texture error,0.052661
26,worst concavity,0.016325
23,worst area,0.012678
13,area error,0.012085
28,worst symmetry,0.010733
24,worst smoothness,0.008452
17,concave points error,0.006261


## Ячейка 9. Борьба с переобучением

### Алгоритм
Ограничиваем глубину дерева.

Этот код демонстрирует один из способов борьбы с переобучением (overfitting) дерева решений путем ограничения его максимальной глубины:

small_tree = DecisionTreeClassifier(max_depth=3, random_state=42):

Мы создаем новую модель дерева решений, но теперь передаем важный гиперпараметр max_depth=3. Это устанавливает жесткое ограничение: дерево не может иметь более 3 уровней ветвления вопросов.
random_state=42 используется для фиксации случайности, чтобы модель обучалась одинаково при каждом запуске.
small_tree.fit(X_train, y_train):

Обучает ограниченное дерево на обучающей выборке. Так как глубина ограничена числом 3, процесс обучения остановится гораздо раньше, не давая модели излишне подстраиваться под мелкие детали и шум в обучающих данных.
small_pred = small_tree.predict(X_test):

Использует обученное ограниченное дерево для предсказания классов на тестовых данных X_test.
small_accuracy = accuracy_score(y_test, small_pred):

Рассчитывает точность (долю правильных ответов) на тестовой выборке.
Почему это важно?
Ограничение глубины (max_depth=3) делает модель более простой. Парадоксально, но более простая модель часто показывает более высокую точность на тестовых данных, чем глубокое дерево. В вашем случае точность выросла с 91.2% (при полной глубине 7) до 93.8% (при глубине 3). Это яркий пример того, как борьба с переобучением улучшает способность модели обобщать правила на новые данные.

In [9]:
small_tree=DecisionTreeClassifier(
    max_depth=3,
    random_state=42
)

small_tree.fit(X_train,y_train)

small_pred=small_tree.predict(X_test)

small_accuracy=accuracy_score(y_test,small_pred)

print("Полное дерево:",accuracy)
print("Ограниченное дерево:",small_accuracy)

assert small_accuracy>0.7

Полное дерево: 0.9122807017543859
Ограниченное дерево: 0.9385964912280702


## Ячейка 10. Мини-проект

### Алгоритм
Формируем отчет.

In [10]:
report={
    "rows":len(X),
    "features":X.shape[1],
    "accuracy":round(accuracy,3),
    "depth":depth,
    "best_feature":importance.iloc[0]["feature"]
}

print(report)

assert report["features"]>10

{'rows': 569, 'features': 30, 'accuracy': 0.912, 'depth': 7, 'best_feature': 'worst radius'}


# GitHub

```bash
git checkout -b feature/block7-lesson07-decision-tree
git add .
git commit -m "feat: add decision tree lesson"
git push -u origin feature/block7-lesson07-decision-tree
```